File: metrics.py
#!/usr/bin/python3
import csv
import subprocess
import os

def get_cgroup_pid_to_job():
    """Returns a mapping of pid to job id from cgroups"""
    pid_to_job = {}

    for subdir, _, files in os.walk("/sys/fs/cgroup"):
        for file in files:
            if file == "cgroup.procs":
                with open(os.path.join(subdir, file), 'r') as f:
                    for pid in f:
                        pid = pid.strip()
                        job_id = subdir.split('/')[-1]
                        pid_to_job[pid] = job_id

    return pid_to_job

def get_nvidia_metrics():
    """Returns gpu metrics as a dictionary from nvidia-smi commands"""
    try:
        # Run nvidia-smi commands and save output to temp files
        subprocess.run('nvidia-smi --query-compute-apps=pid,used_gpu_memory,gpu_name,gpu_uuid --format=csv > compute_apps_usage.csv', shell=True, check=True)
        subprocess.run('nvidia-smi --query-gpu=gpu_uuid,name,utilization.gpu --format=csv > gpu_usage.csv', shell=True, check=True)
    except subprocess.CalledProcessError as e:
        print(f"Error running nvidia-smi commands: {e}")
        return {}

    metrics = {}
    try:
        # Read compute_apps_usage.csv
        with open('compute_apps_usage.csv', 'r') as f:
            reader = csv.reader(f)
            header = next(reader)
            for row in reader:
                pid = row[0]
                gpu_memory_usage = row[1]
                gpu_name = row[2]
                gpu_uuid = row[3]

                metrics[pid] = {
                    'gpu_memory_usage': gpu_memory_usage,
                    'gpu_name': gpu_name,
                    'gpu_uuid': gpu_uuid,
                    'gpu_utilization': 'N/A',  # Initialize with 'N/A'
                }

        # Read gpu_usage.csv and match UUIDs
        with open('gpu_usage.csv', 'r') as f:
            reader = csv.reader(f)
            header = next(reader)
            for row in reader:
                gpu_uuid = row[0]
                gpu_utilization = row[2]

                for pid, metric in metrics.items():
                    if metric['gpu_uuid'] == gpu_uuid:
                        metric['gpu_utilization'] = gpu_utilization

    except IOError as e:
        print(f"Error reading CSV files: {e}")

    return metrics

def get_io_metrics(pid_to_job):
    """Returns io metrics as a dictionary for each pid"""
    io_metrics = {}
    for pid in pid_to_job.keys():
        try:
            with open(f'/proc/{pid}/io', 'r') as f:
                io_data = f.read()
            io_metrics[pid] = parse_io_data(io_data)
        except IOError:
            io_metrics[pid] = {
                'read_bytes': 'N/A',
                'write_bytes': 'N/A'
            }
    return io_metrics

def parse_io_data(io_data):
    """Parse the io data from /proc/[pid]/io"""
    lines = io_data.split('\n')
    io_info = {}
    for line in lines:
        if 'read_bytes' in line or 'write_bytes' in line:
            key, value = line.split(':')
            io_info[key.strip()] = value.strip()
    return io_info
    
def write_to_textfile_collector(pid_to_job, metrics):
    """Writes gpu metrics to the textfile collector format"""
    job_gpu_utilization = {}  # Dictionary to store job-wise GPU utilization
    job_gpu_memory_usage = {}  # Dictionary to store job-wise GPU memory usage
    
    # Collect GPU utilization and memory usage per job_id
    for pid, job in pid_to_job.items():
        metric = metrics.get(pid, {})
        gpu_uuid = metric.get('gpu_uuid', '')
        gpu_utilization = metric.get('gpu_utilization', 'N/A')
        gpu_memory_usage = metric.get('gpu_memory_usage', 'N/A')
        
        if gpu_utilization != 'N/A':
            job_gpu_utilization.setdefault(job, []).append((gpu_uuid, gpu_utilization))
        if gpu_memory_usage != 'N/A':
            job_gpu_memory_usage.setdefault(job, []).append((gpu_uuid, gpu_memory_usage))
    
    # Write the GPU utilization and memory usage to the textfile collector
    with open('/var/lib/node_exporter/textfile_collector/gpu_metrics.prom', 'w') as f:
        for job, gpu_utilizations in job_gpu_utilization.items():
            for gpu_uuid, gpu_utilization in gpu_utilizations:
                f.write(f'cgroups_nvidia_gpu_utilization{{gpu_id="{gpu_uuid}", job_id="{job}"}} {gpu_utilization}\n')
        
        for job, gpu_memory_usages in job_gpu_memory_usage.items():
            for gpu_uuid, gpu_memory_usage in gpu_memory_usages:
                f.write(f'cgroups_nvidia_gpu_memory_usage_in_bytes{{gpu_id="{gpu_uuid}", job_id="{job}"}} {gpu_memory_usage}\n')
              
    # Write the IO metrics to the textfile collector
    with open('/var/lib/node_exporter/textfile_collector/io_metrics.prom', 'a') as f:
        for pid, job in pid_to_job.items():
            io_metric = io_metric.get(pid, {})
            read_bytes = io_metric.get('read_bytes', 'N/A')
            write_bytes = io_metric.get('write_bytes', 'N/A')

            f.write(f'cgroups_io_read_bytes{{pid="{pid}", job_id="{job}"}} {read_bytes}\n')
            f.write(f'cgroups_io_write_bytes{{pid="{pid}", job_id="{job}"}} {write_bytes}\n')


def main():
    pid_to_job = get_cgroup_pid_to_job()
    metrics = get_nvidia_metrics()
    io_metrics = get_io_metrics(pid_to_job)
    write_to_textfile_collector(pid_to_job, metrics, io_metrics)

if __name__ == "__main__":
    main()

File: metrics.go
package main

import (
	"bufio"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var (
	gpuUtilizationMetric = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "gpu_utilization",
		Help: "GPU utilization percentage.",
	}, []string{"gpu_id", "job_id"})

	gpuMemoryUsageMetric = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "gpu_memory_usage_bytes",
		Help: "GPU memory usage in bytes.",
	}, []string{"gpu_id", "job_id"})

	ioReadBytesMetric = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "io_read_bytes",
		Help: "IO read bytes.",
	}, []string{"pid", "job_id"})

	ioWriteBytesMetric = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "io_write_bytes",
		Help: "IO write bytes.",
	}, []string{"pid", "job_id"})
)

func init() {
	// Register the custom metrics with Prometheus's default registry
	prometheus.MustRegister(gpuUtilizationMetric)
	prometheus.MustRegister(gpuMemoryUsageMetric)
	prometheus.MustRegister(ioReadBytesMetric)
	prometheus.MustRegister(ioWriteBytesMetric)

	//Initalize metrics to zero for trakcing 0% utilization
	gpuMemoryUsageMetric.WithLabelValues("none", "none").Set(0)
	gpuUtilizationMetric.WithLabelValues("none", "none").Set(0)
	ioReadBytesMetric.WithLabelValues("none", "none").Set(0)
	ioWriteBytesMetric.WithLabelValues("none", "none").Set(0)
}

//Added function to get current runing jobs

// getJobIDFromPID finds the job ID for a given PID from the Slurm cgroup directory
func getJobIDFromPID(pid string) (string, error) {
	// Base path to Slurm
	basePath := "/sys/fs/cgroup/cpu/slurm"

	// Open the base path directory
	baseDir, err := os.Open(basePath)
	if err != nil {
		return "", fmt.Errorf("failed to open the base directory: %v", err)
	}
	defer baseDir.Close()

	// Read entries
	entries, err := baseDir.Readdirnames(-1)
	if err != nil {
		return "", fmt.Errorf("failed to read the entries in the directory: %v", err)
	}

	// Iterate over each entry looking for uid directories
	for _, entry := range entries {
		if strings.HasPrefix(entry, "uid_") {
			// Construct path for this uid directory
			uidPath := fmt.Sprintf("%s/%s", basePath, entry)

			// Open the uid directory
			uidDir, err := os.Open(uidPath)
			if err != nil {
				continue // If unable to open, skip to next uid directory
			}

			// Read job entries in the uid directory
			jobEntries, err := uidDir.Readdirnames(-1)
			uidDir.Close()
			if err != nil {
				continue // If unable to read, skip to next uid directory
			}

			// Iterate over job entries
			for _, jobEntry := range jobEntries {
				if strings.HasPrefix(jobEntry, "job_") {
					// Construct path for this job directory
					jobPath := fmt.Sprintf("%s/%s/cgroup.procs", uidPath, jobEntry)

					// Attempt to open the cgroup.procs file within this job directory
					file, err := os.Open(jobPath)
					if err != nil {
						continue // If unable to open, skip to next job directory
					}

					// Scan through the cgroup.procs file
					scanner := bufio.NewScanner(file)
					for scanner.Scan() {
						line := scanner.Text()
						if line == pid {
							file.Close()
							return strings.TrimPrefix(jobEntry, "job_"), nil
						}
					}
					file.Close()

					if err := scanner.Err(); err != nil {
						return "", fmt.Errorf("error scanning cgroup file for PID %s in %s: %v", pid, jobPath, err)
					}
				}
			}
		}
	}

	return "", fmt.Errorf("job ID not found for PID %s", pid)
}

func collectGPUMetrics() {
	// Run nvidia-smi to get GPU usage and application memory usage
	gpuInfoCmd := exec.Command("bash", "-c", "nvidia-smi --query-gpu=gpu_uuid,index,name,utilization.gpu --format=csv,noheader")
	gpuInfoOutput, err := gpuInfoCmd.Output()
	if err != nil {
		fmt.Printf("Failed to execute command: %s\n", err)
		return
	}

	computeAppsCmd := exec.Command("bash", "-c", "nvidia-smi --query-compute-apps=pid,used_gpu_memory,gpu_uuid --format=csv,noheader")
	computeAppsOutput, err := computeAppsCmd.Output()
	if err != nil {
		fmt.Printf("Failed to execute command: %s\n", err)
		return
	}

	// Process gpuInfoOutput to map UUID to GPU ID
	gpuInfoLines := strings.Split(strings.TrimSpace(string(gpuInfoOutput)), "\n")
	gpuUUIDToIndex := make(map[string]string)
	for _, line := range gpuInfoLines {
		parts := strings.Split(line, ", ")
		if len(parts) == 4 {
			uuid := parts[0]
			index := parts[1]
			gpuUUIDToIndex[uuid] = index
		}
	}

	// Process computeAppsOutput and update Prometheus metrics
	computeAppsLines := strings.Split(strings.TrimSpace(string(computeAppsOutput)), "\n")
	for _, line := range computeAppsLines {
		parts := strings.Split(line, ", ")
		if len(parts) == 3 {
			pid := parts[0]
			usedMemory, err := strconv.ParseFloat(strings.Trim(parts[1], " MiB"), 64)
			if err != nil {
				fmt.Printf("Error parsing used GPU memory for PID %s: %v\n", pid, err)
				continue
			}
			uuid := parts[2]

			if index, exists := gpuUUIDToIndex[uuid]; exists {
				jobID, err := getJobIDFromPID(pid)
				if err != nil {
					fmt.Printf("Error fetching job ID for PID %s: %v\n", pid, err)
					continue
				}

				gpuMemoryUsageMetric.With(prometheus.Labels{"gpu_id": index, "job_id": jobID}).Set(usedMemory * 1024 * 1024) // Convert MiB to bytes
			}
		}
	}
}

func collectIOMetrics() {
	// Base path to Slurm
	basePath := "/sys/fs/cgroup/cpu/slurm"

	// Open the base path directory
	baseDir, err := os.Open(basePath)
	if err != nil {
		fmt.Printf("Failed to open the base directory: %s\n", err)
		return
	}
	defer baseDir.Close()

	// Read entries
	entries, err := baseDir.Readdirnames(-1)
	if err != nil {
		fmt.Printf("Failed to read the entries in the directory: %s\n", err)
		return
	}

	// Iterate over each entry looking for uid directories
	for _, entry := range entries {
		if strings.HasPrefix(entry, "uid_") {
			// Construct path for this uid directory
			uidPath := fmt.Sprintf("%s/%s", basePath, entry)

			// Open the uid directory
			uidDir, err := os.Open(uidPath)
			if err != nil {
				fmt.Printf("Failed to open UID directory %s: %s\n", uidPath, err)
				continue // If unable to open, skip to next uid directory
			}

			// Read job entries in the uid directory
			jobEntries, err := uidDir.Readdirnames(-1)
			uidDir.Close()
			if err != nil {
				fmt.Printf("Failed to read job entries in UID directory %s: %s\n", uidPath, err)
				continue // If unable to read, skip to next uid directory
			}

			// Iterate over job entries
			for _, jobEntry := range jobEntries {
				if strings.HasPrefix(jobEntry, "job_") {
					// Construct path for this job directory
					jobPath := fmt.Sprintf("%s/%s", uidPath, jobEntry)
					cgroupProcsPath := filepath.Join(jobPath, "cgroup.procs")

					// Check if cgroup.procs file exists
					if _, err := os.Stat(cgroupProcsPath); os.IsNotExist(err) {
						fmt.Printf("No cgroup.procs file for job %s (UID %s), skipping\n", jobEntry, entry)
						continue
					}

					// Read the PIDs from the cgroup.procs file
					pids, err := os.ReadFile(cgroupProcsPath)
					if err != nil {
						fmt.Printf("Failed to read cgroup.procs for job %s (UID %s): %v\n", jobEntry, entry, err)
						continue
					}

					// If no PIDs, initialize I/O metrics to zero and skip this job
					if len(strings.Fields(string(pids))) == 0 {
						fmt.Printf("No PIDs found in cgroup.procs for job %s (UID %s), initializing I/O metrics to zero\n", jobEntry, entry)
						ioReadBytesMetric.With(prometheus.Labels{"pid": "none", "job_id": jobEntry}).Set(0)
						ioWriteBytesMetric.With(prometheus.Labels{"pid": "none", "job_id": jobEntry}).Set(0)
						continue
					}

					// Collect I/O metrics for each PID
					for _, pid := range strings.Fields(string(pids)) {
						ioFilePath := fmt.Sprintf("/proc/%s/io", pid)
						content, err := os.ReadFile(ioFilePath)
						if err != nil {
							fmt.Printf("Error reading IO file for PID %s: %v\n", pid, err)
							continue
						}

						for _, line := range strings.Split(string(content), "\n") {
							parts := strings.Split(line, ":")
							if len(parts) == 2 {
								key := strings.TrimSpace(parts[0])
								value, err := strconv.ParseFloat(strings.TrimSpace(parts[1]), 64)
								if err != nil {
									fmt.Printf("Error parsing IO metric for PID %s: %v\n", pid, err)
									continue
								}

								if key == "read_bytes" {
									ioReadBytesMetric.With(prometheus.Labels{"pid": pid, "job_id": jobEntry}).Set(value)
								} else if key == "write_bytes" {
									ioWriteBytesMetric.With(prometheus.Labels{"pid": pid, "job_id": jobEntry}).Set(value)
								}
							}
						}
					}
				}
			}
		}
	}
}

func main() {
	go func() {
		ticker := time.NewTicker(10 * time.Second)
		for {
			select {
			case <-ticker.C:
				collectGPUMetrics()
				collectIOMetrics()
			}
		}
	}()

	http.Handle("/metrics", promhttp.Handler())
	fmt.Println("Serving metrics at /metrics")
	//Exposes metrics via http://localhost:9060/metrics
	http.ListenAndServe(":9060", nil)
}
